src/riak_kv_wm_timeseries_query.erl

```diff
@@ -54,69 +55,82 @@
-init(Props) ->
-    {ok, #ctx{prefix = proplists:get_value(prefix, Props),
-              riak = proplists:get_value(riak, Props)}}.
-
--spec service_available(#wm_reqdata{}, #ctx{}) ->
-    {boolean(), #wm_reqdata{}, #ctx{}}.
+init(_Props) ->
+    {ok, #ctx{}}.
+
+-spec service_available(#wm_reqdata{}, #ctx{}) -> cb_rv_spec(boolean()).
 %% @doc Determine whether or not a connection to Riak
-%%      can be established.  This function also takes this
-%%      opportunity to extract the 'bucket' and 'key' path
-%%      bindings from the dispatch, as well as any vtag
-%%      query parameter.
-service_available(RD, Ctx = #ctx{riak = RiakProps}) ->
-    case riak_kv_wm_utils:get_riak_client(
-           RiakProps, riak_kv_wm_utils:get_client_id(RD)) of
-        {ok, C} ->
-            {true, RD,
-             Ctx#ctx{api_version = wrq:path_info(api_version, RD),
-                     method = wrq:method(RD),
-                     client = C
-                    }};
-        {error, Reason} ->
-            handle_error({riak_client_error, Reason}, RD, Ctx)
+%%      can be established.
+service_available(RD, Ctx) ->
+    case init:get_status() of
+        {started, _} ->
+            {true, RD, Ctx};
+        Status ->
+            Resp = riak_kv_wm_ts_util:set_error_message("Unable to connect to Riak: ~p",
+                                                        [Status], RD),
+            {false, Resp, Ctx}
     end.
 
+-spec allowed_methods(#wm_reqdata{}, #ctx{}) -> cb_rv_spec([atom()]).
+allowed_methods(RD, Ctx) ->
+    {['POST'], RD, Ctx}.
 
-is_authorized(RD, Ctx) ->
-    case riak_api_web_security:is_authorized(RD) of
-        false ->
-            {"Basic realm=\"Riak\"", RD, Ctx};
-        {true, SecContext} ->
-            {true, RD, Ctx#ctx{security = SecContext}};
-        insecure ->
-            %% XXX 301 may be more appropriate here, but since the http and
-            %% https port are different and configurable, it is hard to figure
-            %% out the redirect URL to serve.
-            handle_error(insecure_connection, RD, Ctx)
+-spec malformed_request(#wm_reqdata{}, #ctx{}) -> cb_rv_spec(boolean()).
+malformed_request(RD, Ctx) ->
+    try
+        {SqlType, SQL} = query_from_request(RD),
+        Table = table_from_sql(SQL),
+        Mod = riak_ql_ddl:make_module_name(Table),
+        {false, RD, Ctx#ctx{sql_type=SqlType,
+                            compiled_query=SQL,
+                            table=Table,
+                            mod=Mod}}
+    catch
+        throw:{query, Reason} ->
+            Response = riak_kv_wm_ts_util:set_error_message("bad query: ~s", [Reason], RD),
+            {true, Response, Ctx};
+        throw:{unsupported_sql_type, Type} ->
+            Response = riak_kv_wm_ts_util:set_error_message(
+                         "The ~p query type is not supported over the HTTP API yet",
+                         [Type], RD),
+            {{halt, 503}, Response, Ctx}
```

Comment:
Exceptions thrown by parser still a bit sad. Perhaps a general catch all error for the non-thrown errors.

```
 curl -X POST http://localhost:8098/ts/1/query -v --data "OHNOES A DANGLING QUOTE ' "
*   Trying ::1...
* connect to ::1 port 8098 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8098 (#0)
> POST /ts/1/query HTTP/1.1
> Host: localhost:8098
> User-Agent: curl/7.43.0
> Accept: */*
> Content-Length: 26
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 26 out of 26 bytes
< HTTP/1.1 500 Internal Server Error
< Server: MochiWeb/1.1 WebMachine/1.10.8 (that head fake, tho)
< Date: Thu, 24 Mar 2016 18:56:16 GMT
< Content-Type: text/html
< Content-Length: 1112
<
<html><head><title>500 Internal Server Error</title></head><body><h1>Internal Server Error</h1>The server encountered an error while processing this request:<br><pre>{error,<<"Unexpected token '''.">>,
    [{riak_ql_lexer,yyaction,4,[{file,"src/riak_ql_lexer.xrl"},{line,131}]},
     {riak_ql_lexer,string,4,
         [{file,
              "/opt/erlang/OTP_R16B02_basho10-rel/lib/parsetools-2.0.10/include/leexinc.hrl"},
          {line,40}]},
     {riak_ql_lexer,lex,1,[{file,"src/riak_ql_lexer.xrl"},{line,157}]},
     {riak_ql_lexer,get_tokens,1,[{file,"src/riak_ql_lexer.xrl"},{line,138}]},
     {riak_kv_wm_timeseries_query,compile_query,1,
         [{file,"src/riak_kv_wm_timeseries_query.erl"},{line,268}]},
     {riak_kv_wm_timeseries_query,malformed_request,2,
         [{file,"src/riak_kv_wm_timeseries_query.erl"},{line,102}]},
     {webmachine_resource,resource_call,3,
         [{file,"src/webmachine_resource.erl"},{line,186}]},
     {webmachine_resource,do,3,
* Connection #0 to host localhost left intact
         [{file,"src/webmachine_resource.erl"},{line,142}]}]}</pre><P><HR><ADDRESS>mochiweb+webmachine web server</ADDRESS></body></html>